src/coprocessor/dag/expr/builtin_compare.rs

```diff
@@ -158,6 +160,112 @@ impl ScalarFunc {
         do_coalesce(self, |v| v.eval_json(ctx, row))
     }
 
+    pub fn greatest_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        do_get_extremum(self, |e| e.eval_int(ctx, row), max)
+    }
+
+    pub fn greatest_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        do_get_extremum(self, |e| e.eval_real(ctx, row), |x, y| x.max(y))
+    }
+
+    pub fn greatest_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        do_get_extremum(self, |e| e.eval_decimal(ctx, row), max)
+    }
+
+    pub fn greatest_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut greatest = mysql::time::zero_datetime(ctx.cfg.tz);
+
+        for exp in &self.children {
+            let s = try_opt!(exp.eval_string_and_decode(ctx, row));
+            match Time::parse_datetime(&s, Time::parse_fsp(&s), ctx.cfg.tz) {
+                Ok(t) => greatest = max(greatest, t),
+                Err(_) => {
+                    if let Err(e) = ctx.handle_invalid_time_error(Error::invalid_time_format(&s)) {
+                        return Err(e);
+                    }
+                }
+            }
+        }
+
+        Ok(Some(Cow::Owned(greatest.to_string().into_bytes())))
+    }
+
+    pub fn greatest_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        do_get_extremum(self, |e| e.eval_string(ctx, row), max)
+    }
+
+    pub fn least_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        do_get_extremum(self, |e| e.eval_int(ctx, row), min)
+    }
+
+    pub fn least_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        do_get_extremum(self, |e| e.eval_real(ctx, row), |x, y| x.min(y))
+    }
+
+    pub fn least_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        do_get_extremum(self, |e| e.eval_decimal(ctx, row), min)
+    }
+
+    pub fn least_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut res = Cow::from(vec![]);
```

Comment:
How about make it `Option<Cow<..>>`, so that you don't need `find_invalid_time` any more, and don't need this vec initialization (just make it `let mut res = None`).

Reply:
Thanks! But IMHO it is better to be consistent with other UDF. How do we return an `Err` without the `Result`?

Reply:
Oh, I mean that `let mut res = Cow::from(vec![]);` can be changed, use None to indicate the state that value has not been set, instead of an allocated (but empty) vector plus a flag. The function signature should keep.